PPM-5 Persistent DB, add BPL UCI interface

itayx opened the pull request:
This PR introduces the UCI interface in the BPL to support the persistent DB feature.
The BPL DB UCI functions use the `libuci` functions to provide a long-term storage system using the UCI system.

vitalybu commented on framework/platform/bpl/include/bpl/bpl_uci.h:
```diff
@@ -0,0 +1,63 @@
+ * See LICENSE file for more details.
+ */
+
+#ifndef _BPL_UCI_H_
```
I don't think this file should be called UCI.
Or at least you should have an interface which then you can implement for UCI.
Need to make sure that we can use other storage backends at a later point.

itayx replied:
Moved it to be internal to the UCI implementation of the BPL_DB

vitalybu commented on framework/platform/bpl/include/bpl/bpl_uci.h:
```diff
@@ -0,0 +1,63 @@
+/******************************** Functions *********************************/
+/****************************************************************************/
+
+bool uci_entry_exists(const std::string &config_file, const std::string &entry_type,
```
Why functions and not a class?

itayx replied:
Currently converting it to a class causes too many design issues with regard to whom would be responsible with handling the instance,
This will be refactored as part of phase 2

itayx replied:
https://jira.prplfoundation.org/browse/PPM-315

vitalybu commented on framework/platform/bpl/include/bpl/bpl_uci.h:
```diff
@@ -0,0 +1,63 @@
+#include <vector>
+
+namespace beerocks {
+namespace bpl {
```
Consider adding another namespace here (maybe `namespace db {`?)

itayx replied:
Added namespace

vitalybu commented on framework/platform/bpl/uci/bpl_uci.cpp:
```diff
@@ -0,0 +1,263 @@
+    if (!(ctx = uci_alloc_context())) {
+        LOG(ERROR) << "uci context allocate failed!";
+        return false;
+    }
```
Since this code repeats in every function, it should probably be extracted into a local module function.
Also consider wrapper it in a `shared_ptr` with a custom destructor to automatically free the context.

Something like this:
```c++
std::shared_ptr<uci_context> alloc_context() {
    auto ctx = std::shared_ptr<uci_context>(uci_alloc_context(), [](uci_context *ctx) {
        if (ctx) {
            uci_free_context(ctx);
        }
    });

    LOG_IF(!ctx, ERROR) << "Failed allocating UCI context!";
    return ctx;
}
```

Then, all you have to do in every function is:
```c++
auto ctx = alloc_context();
if (!ctx) {
    return false;
}
```

And no need to worry about freeing the context.

itayx replied:
Added `shared_ptr` for the context

vitalybu commented on framework/platform/bpl/uci/bpl_uci.cpp:
```diff
@@ -0,0 +1,263 @@
+        return false;
+    }
+
+    struct uci_package *pkg = nullptr;
```
Since we're in C++, you don't have to add the `struct` keyword.

vitalybu commented on framework/platform/bpl/uci/bpl_uci.cpp:
```diff
@@ -0,0 +1,263 @@
+    }
+
+    uci_delete(ctx, ptr);
+
```
Missing `uci_free_context(ctx);` in a lot of functions.
See my `shared_ptr` comment to remove this dependency at all.

itayx replied:
Added `shared_ptr` for the context

VladyslavTupikin reviewed (commented):
A lot of issues with compiling for RAX40 board.

kantera800 commented on framework/platform/bpl/include/bpl/bpl_uci.h:
```diff
@@ -0,0 +1,63 @@
+bool uci_get_entry_type(const std::string &config_file, const std::string &entry_name,
+                        std::string &entry_type);
+
+bool uci_get_param(const std::string &config_file, const std::string &entry_type,
```
what is the difference between entry and param ?
please add doxigen for all interfaces

itayx replied:
Added description to all APIs

kantera800 commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -17,45 +18,94 @@
+    return uci_get_entry(DB_FILE, entry_type, entry_name, params);
 }
 
 bool db_get_entries_by_type(
     const std::string &entry_type,
```
both entry name and type are strings is this on purpose?

itayx replied:
Yes, I don't want to force the user to use a preset amount of types.
This way each place that uses the UCI persistent DB can use their own types

kantera800 commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -17,45 +18,94 @@
+    return uci_set_entry(DB_FILE, entry_type, entry_name, params);
 }
 
 bool db_get_entry(const std::string &entry_type, const std::string &entry_name,
```
shouldn't it be `db_get_entry_by_name`

itayx replied:
No since you either get a single entry by it's unique identifier (name)
or get all entries with the same type

adam1985d replied:
@itayx this is not true
if the API is called with entry_type **set** and entry_name **not-set** and params **set**
you will get the params from all clients - what is a valid use-case for this?

you need to verify that the name is **always** provided:
```
if (entry_name.empty()) {
    LOG(ERROR)
    return false;
}
```
(in the bpl_db layer you need to think from a bpl_db perspective and not from UCI perspective)

itayx committed:
framework: bpl: uci: add new UCI DB interface

preparatory commit.

To support the persistent DB feature, we use the UCI library APIs, and
with it provide a simpler, more straightforward interface.

Added BPL UCI stubs to be used by the BPL DB to read/write/modify the
UCI DB files.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: uci: add UCI helper functions.

preparatory commit.

Added BPL UCI helper functions,
- alloc_context which allocated a uci_context "smart" pointer.
- uci_get_error which returns the output error from the UCI.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: uci: Implement section_exists function.

The implementation of bpl_db_uci function utilizes the `libuci`
functions that provides persistency of the configured data over UCI.

Implemented the section_exists API, the functions checks whether a
section exists in the given configuration.
Existence is determined by matching section name & type (if provided).

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: uci: Implement add_section function.

The implementation of bpl_db_uci function utilizes the `libuci`
functions that provides persistency of the configured data over UCI.

Implemented the add_section API, the functions creates a new named
section with the provided type in the given configuration.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: uci: Implement set_section function.

The implementation of bpl_db_uci function utilizes the `libuci`
functions that provides persistency of the configured data over UCI.

Implemented the set_section API, the function sets the provided options
as described below:
- if the option is provided without a value it is removed from the
  section.
- if the option is provided with value:
-- if it doesn't exist in the section it will be added with its value
-- if it already exist - its value is overwritten
Existing options in the section, that are not provided as part of the
options map will not be changed.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: uci: Implement get_section function.

The implementation of bpl_db_uci function utilizes the `libuci`
functions that provides persistency of the configured data over UCI.

Implemented the get_section API, the functions retrieves all available
(configured) options of the found section.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: uci: Implement get_section_type function.

The implementation of bpl_db_uci function utilizes the `libuci`
functions that provides persistency of the configured data over UCI.

Implemented the get_section_type API, the function retrieves the type of
the found section.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: uci: Implement get_option function.

The implementation of bpl_db_uci function utilizes the `libuci`
functions that provides persistency of the configured data over UCI.

Implemented the get_option API, the function retrieves the value of the
requested option.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: uci: Implement delete_section function.

The implementation of bpl_db_uci function utilizes the `libuci`
functions that provides persistency of the configured data over UCI.

Implemented the delete_section API, the function searches and removes
the found section.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: uci: Implement get_all_sections function.

The implementation of bpl_db_uci function utilizes the `libuci`
functions that provides persistency of the configured data over UCI.

Implemented the get_all_sections API, the function returns all the
available sections names, if the type is specifiers only the sections
with matching types are returned.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

itayx committed:
framework: bpl: use the UCI APIs in the BPL DB.

Utilize the newly implemented bpl_db_uci APIs in the bpl_db.

The following functionality in the bpl_db is implemented:
- db_has_entry - Check whether an entry exists in the DB.
- db_add_entry - Adds a new entry to the DB & fills it with the provided
    parameters, if entry exists (even under different type) the function
    will fail.
- db_set_entry - Set provided parameters for a given entry. If entry
    type is provided the entry must be of the same type.
- db_get_entry - retrieve the matching entry's parameters from the DB.
- db_get_entries_by_type - retrieve all entries with their values for a
    given entry type from the DB. If the type is not provided, retrieve
    all available entries.
- db_remove_entry - remove an entry from the DB. Note that if an
    entry-type is provided, and entry type doesn't match the provided
    entry type, it will not be removed.
Removed the unneeded cppcheck issues that were added before the
implementation.

Relates to JIRA PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

arnout commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -0,0 +1,93 @@
+namespace db {
+
+bool uci_entry_exists(const std::string &config_file, const std::string &entry_type,
+                      const std::string &entry_name)
```
(nitpick) The [proper terminology](https://openwrt.org/docs/guide-user/base-system/uci#uci_dataobject_model) is
```suggestion
bool uci_section_exists(const std::string &config, const std::string &section_type,
                      const std::string &section_name)
```
At least, I think that is what you mean with an "entry"... You may also mean "option" instead of "section".

Oh, and to make things confusing, sometime "package" is used instead of "config" :-) Maybe config_file isn't so bad after all...

itayx replied:
Renamed to match UCI names

arnout commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -0,0 +1,93 @@
+    return true;
+}
+
+bool uci_add_entry(const std::string &config_file, const std::string &entry_type,
```
I don't think it makes sense to have this. Yes, uci.h has a separate uci_add_section function, but as a user of this library I will probably just want to do uci_set_entry() without worrying about adding the section if it doesn't exist yet.

itayx replied:
Yes, that's why in the upper layer (bpl_db) if the entry does not exists we add it.
But I believe that the "set" function by itself should not add a section if it does not exists

But you raised a good point and I reworked the "add" function to simply use "uci_set" instead of using "uci_add_section" & "uci_rename".

itayx replied:
And this also will add the functionally to change the type of the section if it already exists

arnout commented on framework/platform/bpl/uci/db/bpl_db_uci.h:
```diff
@@ -0,0 +1,134 @@
+ * @return true if entry exists in db, false otherwise
+ */
+bool uci_entry_exists(const std::string &config_file, const std::string &entry_type,
+                      const std::string &entry_name);
```
uci.h itself has a uci_context which has to be passed in each call. Therefore, it would make sense to put all this in a class that carries the uci_context (and that will take care of loading and saving).

Also, as @mariomaz will tell you, this class should be split into two classes: an interface class that only has pure virtual methods, and an implementation class that overrides them. That will make sure you can write a unit test for the peristent DB without having an actual UCI.

itayx replied:
While I agree with your suggestion to create a class instance to handle the UCI interactions, we wanted to provide a quick solution to be used for the persistent DB feature and we plan to later-on refactor the BPL UCI interface.

https://jira.prplfoundation.org/browse/PPM-315

arnout commented on framework/platform/bpl/uci/db/bpl_db_uci.h:
```diff
@@ -0,0 +1,134 @@
+ * @brief Set values in entry, updating as needed.
+ * A parameter set to an empty string will remove it from the entry.
+ * A parameter that does not exist will be added to the entry.
+ * A parameter that already exists will be overridden with the new value
```
```suggestion
 * @brief Set values in entry, updating as needed.
 *
 * An option set to an empty string will remove it from the section.
 * An option that does not exist will be added to the section.
 * An option that already exists will be overridden with the new value
 * Options that already exist and are not mentioned in @a params will remain unchanged.
```

I.e. param -> option and entry -> section. Same in other places.

itayx replied:
Renamed to match UCI names

arnout reviewed (approved):
I assume the intention is to squash everything in a single commit?

simonbaren commented on framework/platform/bpl/uci/db/bpl_db_uci.h:
```diff
@@ -0,0 +1,135 @@
+ * @param[in] package_name name of the requested configuration file.
+ * @param[in] section_type type of the requested section.
+ * @param[in] section_name name of the requested section.
+ * @param[in/out] params unordered_map empty initalized map, that will contain found parameters. 
```
All those `in/out` parameters are either in or out. On `set` apis it acts as `in` and on `get`  it's an `out`

itayx replied:
These `in/out` need to be initialized by the caller, therefore they are `in/out`

simonbaren replied:
https://stackoverflow.com/a/6900223   
`IN/OUT: An IN/OUT parameter usually indicates something where both the input and output values are meaningful. For instance, if you have a library function that fills a buffer, it may require you to pass a pointer to the buffer, along with another pointer indicating the length of the buffer. When the function returns, the second pointer may contain the actual number of bytes that have been written to the buffer.`

adam1985d replied:
the map discussed in this thread is [in/out] since the caller can provide specific options to retrieve - if the map is empty the all configured options for that client are retrieved.

simonbaren commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -3,10 +3,40 @@
+{
+    char *err_buf;
+    uci_get_errorstr(ctx, &err_buf, "");
+    return std::string(err_buf);
```
in case `err_buf` is `nullptr` you will cause an exception.   
you can see proof here:   https://repl.it/repls/ShabbyBurlyWorkers#main.cpp   
check for null

itayx replied:
Added a check, and in case `uci_get_errorstr` does not set `err_buf` it'll return an empty string.

simonbaren commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -44,7 +44,48 @@ bool uci_section_exists(const std::string &package_name, const std::string &sect
+    }
+
+    char path[MAX_UCI_BUF_LEN] = {0};
+    if (sprintf(path, package_path, package_name.c_str()) <= 0) {
```
use `snprintf` as it's safer

simonbaren replied:
or better yet, create helper functions to compose those strings instead of c APIs

itayx replied:
Changed throughout

simonbaren commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -144,6 +144,69 @@ bool uci_set_section(const std::string &package_name, const std::string &section
+        if (uci_set(ctx.get(), &opt_ptr) != UCI_OK) {
+            LOG(ERROR) << "Failed to set option " << option.first;
+            return false;
+        } else {
```
no need for else. the if returns from the function anyway

itayx replied:
Removed the else

simonbaren commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -389,6 +389,34 @@ bool uci_get_all_sections(const std::string &package_name, const std::string &se
+        return false;
+    }
+
+    struct uci_ptr ptr;
```
just notice here but I guess it's relevant to all other commits - you can drop the `struct` keyword

itayx replied:
Removed the unneeded `struct` keyword

simonbaren commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -17,45 +19,100 @@
+            LOG(ERROR) << "Failed to get entry " << entry_name;
+            return false;
+        }
+        nested_params[entry_name] = params;
```
```suggestion
        nested_params[entry_name] = std::move(params);
```
more efficient- see complexity @ https://en.cppreference.com/w/cpp/container/unordered_map/operator%3D

itayx replied:
Changed

simonbaren reviewed (approved):
Nothing critical

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.h:
```diff
@@ -0,0 +1,135 @@
+ * @param[in] package_name name of the requested configuration file.
+ * @param[in] section_type type of the requested section.
+ * @param[in] section_name name of the requested section.
+ * @param[in/out] params unordered_map empty initalized map, that will contain found parameters. 
```
```suggestion
 * @param[in/out] options unordered map of parameters to read/get. If the map is empty, all configured parameters will be retrieved.
```

itayx replied:
Updated Doxygen

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.h:
```diff
@@ -0,0 +1,135 @@
+ * @param[in] package_name name of the requested configuration file.
+ * @param[in] section_type type of the requested section.
+ * @param[in] section_name name of the requested section.
+ * @param[in] params unordered_map containing the params the requested section, to be set.
```
```suggestion
 * @param[in] options unordered_map containing a key/value pair of parameters to set.
```

you might want to review rest of Doxygen here if you choose to accept this suggestion

itayx replied:
Updated Doxygens throughout the file

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.h:
```diff
@@ -0,0 +1,135 @@
+ */
+bool uci_set_section(const std::string &package_name, const std::string &section_type,
+                     const std::string &section_name,
+                     const std::unordered_map<std::string, std::string> &options);
```
if you provide at the begining of the header file:
`using OptionsUnorderedMap = std::unordered_map<std::string, std::string>;`
you can change all locations like this:
```suggestion
bool uci_set_section(const std::string &package_name, const std::string &section_type,
                     const std::string &section_name,
                     const OptionsUnorderedMap &options);
```

(I found it total in this PR there are 9 places that can be changed)

(nitpick)

itayx replied:
Added Alias-Declaration for `OptionsUnorderedMap`

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.h:
```diff
@@ -0,0 +1,135 @@
+/**
+ * @brief Get all entries with the same type.
+ * 
+ * @param[in] package_name name of the requested configuration file.
+ * @param[in] section_type type of the requested section. (If empty get sections of all types)
```
(nitpick)
as @morantr said before, for const & not need to write [in] as it is obvious.

itayx replied:
Even if it's a `const &` argument I still think the `[in]` description is good to have if not for clarity then at least for consistency

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -44,7 +44,48 @@ bool uci_section_exists(const std::string &package_name, const std::string &sect
                << section_name;
 
-    return true;
+    auto ctx = alloc_context();
```
suggested commit message:
```
framework: bpl: uci: Implement section_exists function

The implementation of bpl_db_uci function utilizes the `libuci` functions that provides persistency of the configured data over UCI.

Implemented the section_exists API that checks whether a section exists in the given
configuration.
Existence is determined by matching section name & type (if provided).

PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>
```
(of course, make sure to align it to the commit guidelines)

itayx replied:
Updated the commit message

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -44,7 +44,48 @@ bool uci_section_exists(const std::string &package_name, const std::string &sect
+    }
+
+    uci_package *pkg = nullptr;
+    if (!(pkg = uci_lookup_package(ctx.get(), path))) {
```
```suggestion
    uci_package *pkg = uci_lookup_package(ctx.get(), path);
    if (!pkg) {
```

itayx replied:
Moved the function call

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -44,7 +44,48 @@ bool uci_section_exists(const std::string &package_name, const std::string &sect
+               << (section_type.empty() || !found ? "!"
+                                                  : std::string(" with type: ") + section_type);
+
+    return false;
```
```suggestion
    bool section_found  = false;
    // Loop through the sections with a matching section name (section_name)
    uci_element *elm = nullptr;
    uci_section *sec = nullptr;
    uci_foreach_element(&pkg->sections, elm)
    {
        uci_section *sec = uci_to_section(elm);
        // uci enforces sections must have unique name
        if (section_name.compare(sec->e.name) == 0) {
            section_found = true;
            break;
        }
    }
    
    if (!section_found) {
        LOG(DEBUG) << "Section not found for " << section_name;
        return false;
    }
    
    if (!section_type.empty() && !(section_type.compare(sec->type) == 0) {
        LOG(WARNING) << "Section type missmatch for " << section_name << ", requested-type=" << section_type << ", actual-type=" << sec->type;
        return false;
    }

    return true;
```

no special print is required for the successful case - since this might be used in many places and successive cases provide no additional info.

itayx replied:
Refactored the function:
- Removed the successful case print
- Untangled the nested conditions

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -95,6 +95,44 @@ bool uci_add_section(const std::string &package_name, const std::string &section
+    if (uci_set(ctx.get(), &ptr) != UCI_OK) {
+        LOG(ERROR) << "Failed to add new section";
+        return false;
+    }
```
I don't like this:
1. what if we have a mac listed as `not-client` for some reason and it has information not relevant to client
2. I think a better approach would be to "if not exist add - otherwise, return false" or "if exists remove and add a new one"
one of the 2 is fine by me - but the API description should specify the approach taken in the header.

itayx replied:
Updated the function, it checks if the section already exists before calling the `uci_set`
according to the "if not exist add - otherwise, return false" logic

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -144,6 +144,69 @@ bool uci_set_section(const std::string &package_name, const std::string &section
     LOG(TRACE) << "uci_set_section() " << package_name << ".(" << section_type << ")"
                << section_name;
 
+    auto ctx = alloc_context();
```
suggested commit message:
```
framework: bpl: uci: Implement set_section function

The implementation of bpl_db_uci function utilizes the `libuci` functions that provide the persistency of the configured data over UCI.

Implemented the set_section API that sets the provided options as described below:
- if the option is provided without value - it is removed from the section.
- if the option is provided with value:
-- if it doesn't exist in the section it will be added with its value
-- if it already exist - its value is overwritten
Existing options in the sections, that are not provided as part of the options map will not be changed.

PPM-5.
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>
```
(of course, make sure to align it to the commit guidelines)

itayx replied:
Updated commit message

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -144,6 +144,69 @@ bool uci_set_section(const std::string &package_name, const std::string &section
+
+    for (auto &option : options) {
+        char opt_path[MAX_UCI_BUF_LEN] = {0};
+        if (snprintf(opt_path, MAX_UCI_BUF_LEN, option_path, package_name.c_str(),
```
```suggestion
        // Generate a uci path to the option we wish to set
        if (snprintf(opt_path, MAX_UCI_BUF_LEN, option_path, package_name.c_str(),
```

itayx replied:
Added comment to add `snprintf`
also, I plan to refactor it later on in PPM-315

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -144,6 +144,69 @@ bool uci_set_section(const std::string &package_name, const std::string &section
+            return false;
+        }
+
+        LOG(TRACE) << "Trying to lookup: " << opt_path;
```
```suggestion
        LOG(DEBUG) << "Trying to lookup: " << opt_path;
```
in general - if you go into rework of the PR - make sure the logging is as we discussed

itayx replied:
Removed the unneeded logs throughout the code

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -144,6 +144,69 @@ bool uci_set_section(const std::string &package_name, const std::string &section
+            LOG(ERROR) << "UCI failed to lookup ptr for path: " << opt_path << std::endl
+                       << uci_get_error(ctx.get());
+            return false;
+        }
```
the lack of functional commenting makes it hard to understand the real flow.

for example - it looks like if the option doesn't exist in the section yet - you are exiting here - but this doesn't align with the function description.

itayx replied:
Added more comments in the code to better explain the flow

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -144,6 +144,69 @@ bool uci_set_section(const std::string &package_name, const std::string &section
+        LOG(TRACE) << "Trying to lookup: " << opt_path;
+
+        uci_ptr opt_ptr;
+        // Get pointer to value
```
```suggestion
        // Get pointer to option
```

adam1985d reviewed (commented):
commenting as didn't complete the review yet - to allow addressing the comments

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -102,6 +102,49 @@ bool uci_add_section(const std::string &package_name, const std::string &section
+                 section_name.c_str(), section_type.c_str()) <= 0) {
+        LOG(ERROR) << "Failed to compose path";
+        return false;
+    }
```
```suggestion
    // Generate section path
    char sec_path[MAX_UCI_BUF_LEN] = {0};
    if (snprintf(sec_path, MAX_UCI_BUF_LEN, section_type_path, package_name.c_str(),
                 section_name.c_str(), section_type.c_str()) <= 0) {
        LOG(ERROR) << "Failed to compose path";
        return false;
    }
```

itayx replied:
Added comments to `snprintf`

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -102,6 +102,49 @@ bool uci_add_section(const std::string &package_name, const std::string &section
+        LOG(ERROR) << "UCI failed to lookup ptr for path: " << sec_path << std::endl
+                   << uci_get_error(ctx.get());
+        return false;
+    }
```
```suggestion
    uci_ptr sec_ptr;
    // Initialize pointer to section from path.
    if (uci_lookup_ptr(ctx.get(), &sec_ptr, sec_path, true) != UCI_OK) {
        LOG(ERROR) << "UCI failed to lookup ptr for path: " << sec_path << std::endl
                   << uci_get_error(ctx.get());
        return false;
    }
```

itayx replied:
Reworded `uci_lookup_ptr` comments

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -102,6 +102,49 @@ bool uci_add_section(const std::string &package_name, const std::string &section
+        return false;
+    }
+
+    if (uci_set(ctx.get(), &sec_ptr) != UCI_OK) {
+        LOG(ERROR) << "Failed to add new section";
+        return false;
+    }
```
```suggestion
    // Add a new section
    if (uci_set(ctx.get(), &sec_ptr) != UCI_OK) {
        LOG(ERROR) << "Failed to add new section";
        return false;
    }
```
the only reason I think this comment is needed is since the uci_set API is used here to perform the add

itayx replied:
Added comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -102,6 +102,49 @@ bool uci_add_section(const std::string &package_name, const std::string &section
+        return false;
+    }
+
+    // Create delta from changes, this does not change the persistent file.
```
👍

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -156,6 +156,75 @@ bool uci_set_section(const std::string &package_name, const std::string &section
+        }
+        // Set the value of option in uci_ptr
+        // when using uci_set, it will store the change in the context
+        opt_ptr.value = strdup(option.second.c_str());
```
from the strdup description in cppreference:
`Returns a pointer to a null-terminated byte string, which is a duplicate of the string pointed to by str1. The returned pointer must be passed to free to avoid a memory leak.`
https://en.cppreference.com/w/c/experimental/dynamic/strdup

adam1985d replied:
also:
`Return value
A pointer to the newly allocated string, or a null pointer if an error occurred.`

what happens here if strdup returns a null pointer? how is this handled by the uci? as error or removal of the option (like in empty string)?

itayx replied:
Added error handling to the `strdup`, and as for the `free` issue, when we use `uci_free_context` it frees all allocated memory within the context.

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -156,6 +156,75 @@ bool uci_set_section(const std::string &package_name, const std::string &section
+    }
+
+    uci_ptr pkg_ptr;
+    // Initialize uci_ptr according to path & validate package existace
```
```suggestion
    // Initialize package pointer from path & validate package existace
```

itayx replied:
Fixed `uci_lookup_ptr` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -236,6 +236,37 @@ bool uci_get_section(const std::string &package_name, const std::string &section
+    }
+
+    uci_ptr sec_ptr;
+    // Initialize uci_ptr according to path & validate section existace
```
```suggestion
    // Initialize section pointer from path & validate section existace
```

itayx replied:
Fixed `uci_lookup_ptr` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.h:
```diff
@@ -0,0 +1,135 @@
+ * 
+ * @param[in] package_name name of the requested configuration file.
+ * @param[in] section_name name of the requested section.
+ * @param[in/out] section_type initalized string, will contain type of found section.
```
```suggestion
 * @param[in/out] section_type will contain type of found section.
```

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -236,6 +236,37 @@ bool uci_get_section(const std::string &package_name, const std::string &section
+
+    // Loop through the option within the found section
+    uci_element *elm = nullptr;
+    uci_foreach_element(&sec_ptr.s->options, elm)
+    {
```
```suggestion
    uci_foreach_element(&sec_ptr.s->options, elm) {
```

itayx replied:
clang-format disagrees

adam1985d replied:
👍

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -236,6 +236,37 @@ bool uci_get_section(const std::string &package_name, const std::string &section
+    uci_foreach_element(&sec_ptr.s->options, elm)
+    {
+        uci_option *opt = uci_to_option(elm);
+        // Only type [UCI_TYPE_STRING] is supported
```
```suggestion
        // Only type UCI_TYPE_STRING is supported
```

itayx replied:
Fixed `uci_lookup_ptr` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -276,6 +276,27 @@ bool uci_get_section_type(const std::string &package_name, const std::string &se
+    }
+
+    uci_ptr sec_ptr;
+    // Initialize uci_ptr according to path & validate section existace
```
```suggestion
    // Initialize section pointer from path & validate section existace
```

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -276,6 +276,27 @@ bool uci_get_section_type(const std::string &package_name, const std::string &se
+        return false;
+    }
+
+    section_type = sec_ptr.s->type;
```
what is the type of `sec_ptr.s->type` - can it be empty? if so might want to handle it or at least provide a debug print stating it is empty (since this is not the general positive case.

itayx replied:
If a section exists it must have a type (we check the section existence as part of the lookup condition)
The type is `char*` which will call `std::string::operator=` for c-string

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -308,6 +308,33 @@ bool uci_get_option(const std::string &package_name, const std::string &section_
     LOG(TRACE) << "uci_get_param() " << package_name << ".(" << section_type << ")" << section_name
                << "." << option_name;
 
+    if (!uci_section_exists(package_name, section_type, section_name)) {
+        LOG(ERROR) << "section " << section_name << " of type " << section_type
+                   << " was not found!";
+        return false;
+    }
```
can it be that the option exists with an empty value? or UCI clears it?

itayx replied:
The UCI can't "hold" an empty value

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -308,6 +308,33 @@ bool uci_get_option(const std::string &package_name, const std::string &section_
+    }
+
+    uci_ptr opt_ptr;
+    // Initialize uci_ptr according to path & validate option existace
```
```suggestion
    // Initialize option pointer from path & validate option existace
```

itayx replied:
Fixed `uci_lookup_ptr` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -308,6 +308,33 @@ bool uci_get_option(const std::string &package_name, const std::string &section_
+    }
+
+    char opt_path[MAX_UCI_BUF_LEN] = {0};
+    if (snprintf(opt_path, MAX_UCI_BUF_LEN, option_path, package_name.c_str(), section_name.c_str(),
```
```suggestion
    // Generate path to the requested option
    char opt_path[MAX_UCI_BUF_LEN] = {0};
    if (snprintf(opt_path, MAX_UCI_BUF_LEN, option_path, package_name.c_str(), section_name.c_str(),
```

itayx replied:
Added `snprintf` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -308,6 +308,33 @@ bool uci_get_option(const std::string &package_name, const std::string &section_
+        return false;
+    }
+
+    option_value = opt_ptr.o->v.string;
```
The same comment as above:
can it be that the option exists with an empty value? or UCI clears it?

if there might be empty value - please add handling or at least debug print

itayx replied:
The UCI does not support empty values,
This is the same reason why all section must have a `type` as the `type` counts as the section's "value"

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -345,6 +345,64 @@ bool uci_delete_section(const std::string &package_name, const std::string &sect
     LOG(TRACE) << "uci_delete_section() " << package_name << ".(" << section_type << ")"
                << section_name;
 
+    if (!uci_section_exists(package_name, section_type, section_name)) {
+        LOG(ERROR) << "section " << section_name << " of type " << section_type
+                   << " was not found!";
+        return false;
+    }
```
```suggestion
    // Verify requested section exists
    if (!uci_section_exists(package_name, section_type, section_name)) {
        LOG(ERROR) << "section " << section_name << " of type " << section_type
                   << " was not found!";
        return false;
    }
```
(nitpick)

itayx replied:
Added comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -345,6 +345,64 @@ bool uci_delete_section(const std::string &package_name, const std::string &sect
+                 section_name.c_str()) <= 0) {
+        LOG(ERROR) << "Failed to compose path";
+        return false;
+    }
```
```suggestion
    // Generate section path
    char sec_path[MAX_UCI_BUF_LEN] = {0};
    if (snprintf(sec_path, MAX_UCI_BUF_LEN, section_path, package_name.c_str(),
                 section_name.c_str()) <= 0) {
        LOG(ERROR) << "Failed to compose path";
        return false;
    }
```

itayx replied:
Added `snprintf` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -345,6 +345,64 @@ bool uci_delete_section(const std::string &package_name, const std::string &sect
+        LOG(ERROR) << "UCI failed to lookup ptr for path: " << sec_path << std::endl
+                   << uci_get_error(ctx.get());
+        return false;
+    }
```
```suggestion
    // Initialize section pointer from path & validate section existace
    if (uci_lookup_ptr(ctx.get(), &sec_ptr, sec_path, true) != UCI_OK || !sec_ptr.s) {
        LOG(ERROR) << "UCI failed to lookup ptr for path: " << sec_path << std::endl
                   << uci_get_error(ctx.get());
        return false;
    }
```

itayx replied:
Fixed `uci_lookup_ptr` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -345,6 +345,64 @@ bool uci_delete_section(const std::string &package_name, const std::string &sect
+
+    uci_ptr sec_ptr;
+    // Initialize uci_ptr according to path & validate section existace
+    if (uci_lookup_ptr(ctx.get(), &sec_ptr, sec_path, true) != UCI_OK || !sec_ptr.s) {
```
also, I see here you verified the section pointer is valid - is this required in all other locations as well?

**this comment is for all code locations that use the `uci_lookup_ptr`**

itayx replied:
Only for places where we want to verify that the element we are looking for exists.
For example in get_section - yes, add_section - no.

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -345,6 +345,64 @@ bool uci_delete_section(const std::string &package_name, const std::string &sect
+    if (snprintf(pkg_path, MAX_UCI_BUF_LEN, package_path, package_name.c_str()) <= 0) {
+        LOG(ERROR) << "Failed to compose path";
+        return false;
+    }
```
```suggestion
    // Generate package path
    char pkg_path[MAX_UCI_BUF_LEN] = {0};
    if (snprintf(pkg_path, MAX_UCI_BUF_LEN, package_path, package_name.c_str()) <= 0) {
        LOG(ERROR) << "Failed to compose path";
        return false;
    }
```

itayx replied:
Added `snprintf` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -345,6 +345,64 @@ bool uci_delete_section(const std::string &package_name, const std::string &sect
+        LOG(ERROR) << "UCI failed to lookup ptr for path: " << pkg_path << std::endl
+                   << uci_get_error(ctx.get());
+        return false;
+    }
```
```suggestion
    // Initialize packge pointer from path & validate package existace
    if (uci_lookup_ptr(ctx.get(), &pkg_ptr, pkg_path, true) != UCI_OK || !pkg_ptr.p) {
        LOG(ERROR) << "UCI failed to lookup ptr for path: " << pkg_path << std::endl
                   << uci_get_error(ctx.get());
        return false;
    }
```

itayx replied:
Fixed `uci_lookup_ptr` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -345,6 +345,64 @@ bool uci_delete_section(const std::string &package_name, const std::string &sect
+    if (uci_save(ctx.get(), pkg_ptr.p) != UCI_OK) {
+        LOG(ERROR) << "Failed to save changes!" << std::endl << uci_get_error(ctx.get());
+        return false;
+    }
```
```suggestion
    // Create delta of changes, this does not change the persistent file.
    if (uci_save(ctx.get(), pkg_ptr.p) != UCI_OK) {
        LOG(ERROR) << "Failed to save changes!" << std::endl << uci_get_error(ctx.get());
        return false;
    }
```

itayx replied:
Changed

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -412,6 +412,35 @@ bool uci_get_all_sections(const std::string &package_name, const std::string &se
+    if (snprintf(pkg_path, MAX_UCI_BUF_LEN, package_path, package_name.c_str()) <= 0) {
+        LOG(ERROR) << "Failed to compose path";
+        return false;
+    }
```
```suggestion
    // Generate package path
    char pkg_path[MAX_UCI_BUF_LEN] = {0};
    if (snprintf(pkg_path, MAX_UCI_BUF_LEN, package_path, package_name.c_str()) <= 0) {
        LOG(ERROR) << "Failed to compose path";
        return false;
    }
```

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -412,6 +412,35 @@ bool uci_get_all_sections(const std::string &package_name, const std::string &se
+        LOG(ERROR) << "UCI lookup failed for path: " << pkg_path << std::endl
+                   << uci_get_error(ctx.get());
+        return false;
+    }
```
```suggestion
    // Initialize packge pointer from path & validate package existace
    if (uci_lookup_ptr(ctx.get(), &pkg_ptr, pkg_path, true) != UCI_OK || !pkg_ptr.p) {
        LOG(ERROR) << "UCI lookup failed for path: " << pkg_path << std::endl
                   << uci_get_error(ctx.get());
        return false;
    }
```

itayx replied:
Fixed `uci_lookup_ptr` comment

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -412,6 +412,35 @@ bool uci_get_all_sections(const std::string &package_name, const std::string &se
+    {
+        uci_section *sec = uci_to_section(elm);
+        // If section type matches add to result
+        if (section_type.empty() || section_type.compare(sec->type) == 0) {
```
👍

adam1985d commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -6,7 +6,9 @@
  */
 
-#include "bpl/bpl_db.h"
+#include <bpl/bpl_db.h>
```
suggested commit message:
```
framework: bpl: use the UCI APIs in the BPL DB.

Utilize the newly implemented bpl_db_uci APIs in the bpl_db.

The following functionality in the bpl_db is implemented:
- db_has_entry - check whether an entry exists in the DB.
- db_add_entry - adds an entry to the DB. If it already exists in the DB, remove before add.
- db_set_entry - set provided parameters for a given entry of a given entry type.
- db_get_entry - retrieve the entry parameters from the DB.
- db_get_entries_by_type - retrieve all entries of a given type from the DB.
- db_remove_entry - remove an entry from the DB. Note that if an entry-type is provided, and entry type doesn't match the provided entry type, it will not be removed.

PPM-5.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>
```

adam1985d replied:
note the following line in the suggested commit message:
- db_add_entry - adds an entry to the DB. **If it already exists in the DB, remove before add.**

Adding client entry should either remove the old entry or return an error since entry already exists.
in this specific case exists doesn't matter the entry-type (since if another type exists it is also handled as exists)

itayx replied:
Updated the commit message

adam1985d commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -17,45 +19,100 @@
+        // params map is empty, getting all parameters
+        return db::uci_get_section(DB_FILE, entry_type, entry_name, params);
+    }
+    for (auto &param : params) {
+        // params map is not empty, getting selected parameters
+        if (!db::uci_get_option(DB_FILE, entry_type, entry_name, param.first, param.second)) {
+            LOG(ERROR) << "Failed to get " << param.first;
+            // TODO:
+            return false;
+        }
+    }
```
see my comment above regarding empty-name

itayx replied:
Added check for `!entry_name.empty()` in all functions and in `db_add_entry` for checked for `!entry_type.empty()` since the add function must have a type.

adam1985d commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -17,45 +19,100 @@
+            LOG(ERROR) << "Failed to get entry " << entry_name;
+            return false;
+        }
+        nested_params[entry_name] = std::move(params);
```
I am not sure this is a valid case (at least not to the clients in DB) but can `params` here be empty and `db::uci_get_section` return true? if can happen - check `params` before `move` - if not, ignore.

itayx replied:
In case the the `params` is empty it'll set an empty nested map under the `entry_name`

adam1985d commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -17,45 +19,100 @@
+    LOG(TRACE) << entry_type << ":" << entry_name;
+
+    if (!db::uci_section_exists(DB_FILE, entry_type, entry_name)) {
+        LOG(DEBUG) << "Entry " << entry_name << " not found!";
```
should print entry type if not empty - since it is a limiter in the "section_exists" check

itayx replied:
Extended the DEBUG print in `db_set_entry` & `db_remove_entry` (places where we additionally call `db::uci_section_exists`).

adam1985d replied:
I don't see you added the entry_type print here

itayx replied:
You're right, didn't push it.. my bad
Fixed

adam1985d commented on ci/cppcheck/cppcheck_existing_issues.txt:
```diff
@@ -183,8 +183,6 @@ framework/platform/bpl/uci/arp/monitor/arp_monitor.cpp: style: Unused variable:
 framework/platform/bpl/uci/cfg/bpl_cfg.cpp: style: The scope of the variable 'retVal' can be reduced. [variableScope]    int retVal = 0;
 framework/platform/bpl/uci/cfg/bpl_cfg.cpp: style: Variable 'retVal' is assigned a value that is never used. [unreadVariable]    int retVal = 0;
 framework/platform/bpl/uci/cfg/bpl_cfg_uci.cpp: style: The scope of the variable 'scanf_res' can be reduced. [variableScope]        int scanf_res;
-framework/platform/bpl/uci/db/bpl_db.cpp: style: Parameter 'nested_params' can be declared with const [constParameter]    std::unordered_map<std::string, std::unordered_map<std::string, std::string>> &nested_params)
-framework/platform/bpl/uci/db/bpl_db.cpp: style: Parameter 'params' can be declared with const [constParameter]                  std::unordered_map<std::string, std::string> &params)
```
this should be squashed into the commit that resolved it and shouldn't be a separate commit.

adam1985d reviewed (changes requested):
I had some comments I want to review before merge although not critical - so not-approving.

adam1985d commented on framework/platform/bpl/uci/db/bpl_db_uci.cpp:
```diff
@@ -158,6 +158,82 @@ bool uci_set_section(const std::string &package_name, const std::string &section
+    }
+
+    // To save the changes made we need to call "uci_commit" with the package
+    // Loading a uci_ptr will also load the changed saved in the context's delta
```
```suggestion
    // Loading a uci_ptr will also load the changes saved in the context's delta
```

itayx replied:
Fixed

adam1985d commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -17,45 +19,121 @@
+    LOG(TRACE) << entry_type << ":" << entry_name;
+
+    if (entry_name.empty()) {
+        LOG(ERROR) << "Entry name must be set";
```
```suggestion
        LOG(ERROR) << "Entry name must be provided";
```

itayx replied:
Fixed

adam1985d commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -17,45 +19,121 @@
+    LOG(TRACE) << entry_type << ":" << entry_name;
+
+    if (entry_name.empty()) {
+        LOG(ERROR) << "Entry name must be set";
```
```suggestion
        LOG(ERROR) << "Entry name must be provided";
```

itayx replied:
Fixed
